Evaluate defaults when adding new features in transaction group

elpaso opened the pull request:
Fixes #37222

elpaso committed:
Evaluate defaults when adding new features in transaction group

Fixes #37222

m-kuhn commented:
This is only required for sqlite based formats. Postgresql supports evaluating default values without an insert. Can we have an isolated patch?

elpaso commented:
@m-kuhn sure, what do you suggest to do? An ugly provider key switch?

m-kuhn commented:
Or fix the default value functions in the gpkg provider to have it fixed at the source?
Or let the provider report capabilities and switch based on them?

elpaso commented:
> Or fix the default value functions in the gpkg provider to have it fixed at the source?

I'm picking for this one, with spatialite support too.

m-kuhn commented:
Good choice :)

elpaso committed:
GPKG and spatialite AUTOINCREMENT: get next value from sequence

for PK default value, fixes #37222

Also, fix dangling transactions for spatialite.

m-kuhn commented on src/core/providers/ogr/qgsogrprovider.cpp:
```diff
@@ -1452,13 +1452,72 @@ QVariant QgsOgrProvider::defaultValue( int fieldId ) const
+  // Get next sequence value for sqlite in case we are inside a transaction
+  if ( mOgrOrigLayer &&
+       mTransaction &&
+       mDefaultValues.value( fieldId, QString() ) == tr( "Autogenerate" ) &&
```
Is "Autogenerate" translated? Isn't that coming from sqlite itself?

elpaso replied:
nope

m-kuhn replied:
fingers crossed we have a consistent translation of this magic value everywhere.

elpaso replied:
it's the same for OGR provider, I think I homogenized spatialite provider taking OGR as the source of truth a while go ... I think it's ok to translate it because it's exposed to the users but yeah: we definitely need to pay attention to translate it all the times we refer to it in the code.

m-kuhn replied:
I agree, let's keep it like this for now.

To make the link: https://github.com/qgis/QGIS/issues/36396

m-kuhn commented on src/core/providers/ogr/qgsogrprovider.cpp:
```diff
@@ -1452,13 +1452,72 @@ QVariant QgsOgrProvider::defaultValue( int fieldId ) const
+
+        if ( ! ok )
+        {
+          QgsMessageLog::logMessage( tr( "Error retrieving next sequence value for %1" ).arg( QString( mOgrOrigLayer->name() ) ), tr( "OGR" ) );
```
```suggestion
          QgsMessageLog::logMessage( tr( "Error retrieving next sequence value for %1" ).arg( QString::fromUtf8( mOgrOrigLayer->name() ) ), tr( "OGR" ) );
```

Does the same, just explicit

m-kuhn commented on src/core/qgssqliteutils.cpp:
```diff
@@ -203,6 +203,49 @@ QSet<QString> QgsSqliteUtils::uniqueFields( sqlite3 *connection, const QString &
+    {
+      if ( dsPtr.exec( QStringLiteral( "UPDATE sqlite_sequence SET seq = %1 WHERE name = %2" )
+                       .arg( QString::number( ++result ) )
+                       .arg( quotedTableName ), errorMessage ) != SQLITE_OK )
```
```suggestion
                       quotedTableName ), errorMessage ) != SQLITE_OK )
```

m-kuhn commented on src/core/qgssqliteutils.cpp:
```diff
@@ -203,6 +203,49 @@ QSet<QString> QgsSqliteUtils::uniqueFields( sqlite3 *connection, const QString &
+    else // increment
+    {
+      if ( dsPtr.exec( QStringLiteral( "UPDATE sqlite_sequence SET seq = %1 WHERE name = %2" )
+                       .arg( QString::number( ++result ) )
```
```suggestion
                       .arg( QString::number( ++result ),
```

Optimization

m-kuhn commented:
Looks good. I don't know the sequence handling of sqlite enough to judge it low level.

elpaso committed:
Apply suggestions from code review

Co-authored-by: Matthias Kuhn <matthias@opengis.ch>

elpaso committed:
Code layout

nyalldawson commented:
This looks potentially dangerous -- can we delay till after 3.14.0?

elpaso commented:
> This looks potentially dangerous -- can we delay till after 3.14.0?

OK, IMO it's not high risk but I'm not in a hurry.
